driver/src/main/java/org/neo4j/driver/v1/Result.java

```diff
@@ -18,29 +18,40 @@
  */
 package org.neo4j.driver.v1;
 
-import org.neo4j.driver.v1.exceptions.ClientException;
+import java.util.List;
+
 
 /**
- * The result of running a statement, a stream of records. The result interface can be used to iterate over all the
- * records in the stream, and for each record to access the fields within it using the {@link #get(int) get} methods.
+ * The result of running a statement, a stream of records represented as a cursor.
+ *
+ * The result cursor can be used to iterate over all the records in the stream and provide access
+ * to their content.
  *
- * Results are valid until the next statement is run or until the end of the current transaction, whichever comes
- * first.
+ * Results are valid until the next statement is run or until the end of the current transaction,
+ * whichever comes first.
  *
- * To keep a result around while further statements are run, or to use a result outside the scope of the current
- * transaction, see {@link #retain()}.
+ * Initially, before {@link #next()} has been called at least once, all field values are null.
+ *
+ * To keep a result around while further statements are run, or to use a result outside the scope
+ * of the current transaction, see {@link #asList()}.
  */
-public interface Result
+public interface Result extends RecordView, Resource
 {
     /**
-     * Retrieve and store the entire result stream. This can be used if you want to
-     * iterate over the stream multiple times or to store the whole result for later use.
+     * Retrieve the zero based position of the cursor in the stream of records.
+     *
+     * Initially, before {@link #next()} has been called at least once, the position is -1.
      *
-     * This cannot be used if you have already started iterating through the stream using {@link #next()}.
+     * @return the current position of the cursor
+     */
+    int position();
+
+    /**
+     * Test if the cursor is positioned at the last stream record of if the stream is empty.
      *
-     * @return {@link ReusableResult}
+     * @return true if the cursor is at the last record or the stream is empty.
      */
-    ReusableResult retain();
+    boolean atEnd();
```

Comment:
`atStart` is covered by `position() == -1`; `atEnd` is the only condition not detectable via `position()` hence we added the extra method. It is used by `single`; without `atEnd()` we couldn't implement that reliably without navigating past the `single` record. Also it may be useful for a future `peek` use case that we discussed.